examples/shader/custom_render_phase.rs

```diff
@@ -0,0 +1,505 @@
+        self.entity
+    }
+
+    #[inline]
+    fn sort_key(&self) -> Self::SortKey {
+        FloatOrd(self.distance)
+    }
+
+    #[inline]
+    fn draw_function(&self) -> DrawFunctionId {
+        self.draw_function
+    }
+
+    #[inline]
+    fn batch_range(&self) -> &Range<u32> {
+        &self.batch_range
+    }
+
+    #[inline]
+    fn batch_range_mut(&mut self) -> &mut Range<u32> {
+        &mut self.batch_range
+    }
+
+    #[inline]
+    fn dynamic_offset(&self) -> Option<NonMaxU32> {
+        self.dynamic_offset
+    }
+
+    #[inline]
+    fn dynamic_offset_mut(&mut self) -> &mut Option<NonMaxU32> {
+        &mut self.dynamic_offset
+    }
+}
+
+impl CachedRenderPipelinePhaseItem for CustomPhaseItem {
+    fn cached_pipeline(&self) -> CachedRenderPipelineId {
+        self.pipeline
+    }
+}
+
+#[derive(Resource)]
+pub struct CustomPipeline {
+    mesh_pipeline: MeshPipeline,
+    bind_group_layout: BindGroupLayout,
+    shader: Handle<Shader>,
+}
+
+impl FromWorld for CustomPipeline {
+    fn from_world(world: &mut World) -> Self {
+        let render_device = world.resource::<RenderDevice>();
+
+        let bind_group_layout = render_device.create_bind_group_layout(
+            "custom_bind_group_layout",
+            &[BindGroupLayoutEntry {
+                binding: 0,
+                ty: BindingType::Buffer {
+                    ty: BufferBindingType::Uniform,
+                    has_dynamic_offset: true,
+                    min_binding_size: Some(CustomMaterial::min_size()),
+                },
+                visibility: ShaderStages::FRAGMENT,
+                count: None,
+            }],
+        );
+
+        let shader = world
+            .resource::<AssetServer>()
+            .load("shaders/custom_draw.wgsl");
+
+        let mesh_pipeline = world.resource::<MeshPipeline>().clone();
+        CustomPipeline {
+            shader,
+            mesh_pipeline,
+            bind_group_layout,
+        }
+    }
+}
+
+// TODO explain what SpecializedMeshPipeline are
+impl SpecializedMeshPipeline for CustomPipeline {
+    type Key = MeshPipelineKey;
+    fn specialize(
+        &self,
+        key: Self::Key,
+        layout: &MeshVertexBufferLayout,
+    ) -> Result<RenderPipelineDescriptor, SpecializedMeshPipelineError> {
+        println!("specialize custom");
+        let mut desc = self.mesh_pipeline.specialize(key, layout)?;
+
+        desc.label = Some("custom_mesh_pipeline".into());
+
+        // The layout of the pipeline
+        // It's important that it matches the order specified in the `DrawCustom`
+        desc.layout = vec![
+            self.mesh_pipeline
+                .get_view_layout(MeshPipelineViewLayoutKey::from(key))
+                .clone(),
+            self.bind_group_layout.clone(),
+            self.mesh_pipeline.mesh_layouts.model_only.clone(),
+        ];
+        desc.vertex.shader = self.shader.clone();
+        desc.fragment.as_mut().unwrap().shader = self.shader.clone();
```

Comment:
why does one require `as_mut().unwrap()` and the other not?

Reply:
The fragment shader is optional but the vertex shader is required. I don't remember what happens if you don't provide a fragment shader but it isn't required